File: daily/2024/1593_Split_a_String_Into_the_Max_Number_of_Unique_Substrings.cpp
#include <iostream>
#include <vector>
#include <string>
#include <unordered_set>

using namespace std;

class Solution {
public:
    int ans;
    void recursion(string& s, int i, string curr, unordered_set<string>& substr){
        if (i == s.length()){
            if (substr.find(curr) == substr.end()){
                ans = max(ans, (int)substr.size() + 1);
            }
            return;
        }
        if (curr != ""){
            if (substr.find(curr) == substr.end()){
                substr.insert(curr);
                recursion(s, i+1, string(1, s[i]), substr);
                substr.erase(curr);
            }
        }
        curr += s[i];
        recursion(s, i+1, curr, substr);
    }
    int maxUniqueSplit(string s) {
        ans = 1;
        unordered_set<string> substr;
        recursion(s, 0, "", substr);
        return ans;
    }
};

int main(){
    string s = "ababccc";
    // string s = "aba";
    // string s = "aa";
    Solution S;
    cout << S.maxUniqueSplit(s) << endl;
    return 0;
}

File: daily_python/1593_Split_a_String_Into_the_Max_Number_of_Unique_Substrings.py
import os
import sys

class Solution:
    def recursion(self, s, i, curr, substr):
        if i == len(s):
            if curr not in substr:
                self.ans = max(self.ans, len(substr) + 1)
            return
        if curr and curr not in substr:
            substr.append(curr)
            self.recursion(s, i+1, s[i], substr)
            substr.pop()
        curr += s[i]
        self.recursion(s, i+1, curr, substr)
        
    def maxUniqueSplit(self, s: str) -> int:
        self.ans = 0
        substr = []
        self.recursion(s, 0, "", substr)
        return self.ans
    
def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            s = lines[0].split('=')[1].strip()[1:-2]
            dataset.append(s)
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for s in dataset:
        results.append(solution.maxUniqueSplit(s))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')